### ISSUE TITLE
retain types on Matchless AST

### ISSUE THREAD

**Issue by johnynek** [johnynek/bosatsu#668]
the types were discarded when building the matchless AST.

To generate monomorphic code, that seems like it was a mistake. We don't need it to generate python, js, or similar untyped dynamic languages, but efficient generation of C, Rust, Go, Java, ... would benefit from having the types.

**Comment by johnynek** [comment]
should be complete before #629

**Comment by johnynek** [comment]
so, this is a bit more complex. See #1749. Currently we actually support polymorphic recursion (well, we did, maybe we don't at the moment, but would like to).

We can do this because we erase the types in matchless so polymorphic recursion is just regular recursion.

If we keep the types, we can't lower into a loop in Matchless because the type is different on every loop through.

This is a solvable problem if we introduce two new nodes: `Box` and `Unbox` which are basically like `T => Any` and `Any => T`. In this way, we would box the types that change in the loop up to `Any` (or `exists a. a` effectively), run the loop, and then unbox at the end of the loop back to the right type.

It's not quite that simple because during the loop any of types that change would need to be boxed originally and then undone.

If this were physical (i.e. if you use the matchless type information to generate a typed lowering of the code, to say generate typed JVM code or something) you'd have to have a way to box and unbox types on functions. On the JVM since everything is boxed actually this would be fine, but if you tried to generate rust (which is monomorphized) this would fail.

I occasionally hit the issue in scala when it couldn't create `@tailrec` functions with polymorphic recursion (because they supported .Net once upon a time, they got rid of that and then allowed polymorphic recursion).

It would be nice to have the types, but also I don't want it to restrict the language to do so.

**Comment by johnynek** [comment]
closing in favor of the design here: https://github.com/johnynek/bosatsu/issues/1788 

no types on Matchless is *feature* not a bug.
